### PR TITLE
fix: Sidebar and ResizablePane more closely track native MacOS behavior

### PR DESCRIPTION
**Pull request by jmatth**
In native applications, resizing the sidebar or a split view tracks the horizontal coordinates of the cursor. Currently in `macos_ui`, resizing is done based on the drag delta on each frame. This leads to what is probably a bug where if you drag past the min or max width and then back, the elements will continue to resize even though your cursor is no longer near the divider. Here's a recording of the current behavior to illustrate:

https://user-images.githubusercontent.com/1316184/150221976-719fbc29-9621-4842-a5c5-5e1fa0d0de28.mov

Also in most native apps, dragging the sidebar far enough to the left will cause it to instantly close, rather than just stop at its minimum width. I implemented this behavior, along with a flag to disable it if desired.

Here's a recording showing the updated behavior from this PR:

https://user-images.githubusercontent.com/1316184/150222338-b83a1dba-3897-419e-befd-f3248df701d5.mov

## Pre-launch Checklist

- [x] I have run `dartfmt` on all changed files 
- [x] I have incremented the package version as appropriate and updated `CHANGELOG.md` with my changes 
- [x] I have added/updated relevant documentation 
- [x] I have run "optimize/organize imports" on all changed files
- [x] I have addressed all analyzer warnings as best I could

### CODE REVIEW

**Comment by cadenkriese** [comment]
It looks like macOS or at least finder has some more complex snapping behavior around the minimum widths. I also noticed the haptics in my trackpad going whenever there was a snap as a minor detail.

https://user-images.githubusercontent.com/13630061/151046455-8429681a-13b1-4de7-b04b-77cccf258d99.mp4

**Comment by jmatth** [comment]
Oh wow, never noticed that before. I'm seeing the following behavior on my machine:

1. There's a particular width the sidebar will snap to when you drag close enough. This snapping is accompanied by haptic feedback from the force touch trackpad
2. Dragging past that point allows you to continue shrinking the sidebar
3. The sidebar stops shrinking entirely at a particular minimum width but does not immediately jump to being closed
4. Once the mouse cursor passes some other invisible cutoff, the sidebar snaps closed

I'm only observing this behavior in the Finder app. Other built-in and 3rd party native applications I tested either have the behavior I implemented here, with some also including the functionality from point 3 above.

Everything but the haptic feedback should be easy enough to add via optional params. Unfortunately I don't see an easy way to add the haptic feedback, as [HapticFeedback](https://api.flutter.dev/flutter/services/HapticFeedback-class.html) only supports iOS and Android right now. It might be possible to make it work via a plugin that provides the appropriate platform channels for MacOS but I don't have any experience with that.

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+## [0.10.2]
+* Sidebar and ResizeablePane more precicely track cursor location
+* Sidebar can now be closed by dragging below its minWidth
+* Sidebar can now be configured to snap into place when dragged near its startWidth
+*
```
Comment (GroovinChip) [inline CHANGELOG.md]:
```suggestion

```

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,8 @@
+## [0.10.2]
+* Sidebar and ResizeablePane more precicely track cursor location
```
Comment (GroovinChip) [inline CHANGELOG.md]:
```suggestion
* Sidebar and ResizablePane more precisely track cursor location
```

File: lib/src/layout/resizable_pane.dart [hunk 3]
```diff
@@ -110,15 +111,19 @@ class _ResizablePaneState extends State<ResizablePane> {
         cursor: _cursor,
         child: const SizedBox(width: 5),
       ),
+      onHorizontalDragStart: (_) {
```
Comment (GroovinChip) [inline lib/src/layout/resizable_pane.dart]: Please add a test for this behavior

File: lib/src/layout/resizable_pane.dart [hunk 4]
```diff
@@ -110,15 +111,19 @@ class _ResizablePaneState extends State<ResizablePane> {
...
+        _dragStartWidth = _width;
+      },
       onHorizontalDragUpdate: (details) {
         setState(() {
+          final newWidth = _resizeOnRight
```
Comment (GroovinChip) [inline lib/src/layout/resizable_pane.dart:121]: Please add a test for this behavior

File: lib/src/layout/sidebar.dart [hunk 5]
```diff
@@ -13,13 +13,16 @@ class Sidebar {
...
+    double? dragClosedBuffer,
+    this.snapToStartBuffer,
     this.maxWidth = 400.0,
     this.startWidth,
     this.padding = EdgeInsets.zero,
     this.windowBreakpoint = 556.0,
     this.bottom,
     this.topOffset = 51.0,
-  });
+  }) : dragClosedBuffer = dragClosedBuffer ?? minWidth / 2;
```
Comment (GroovinChip) [inline lib/src/layout/sidebar.dart:25]: Please add a test for this behavior

File: lib/src/layout/sidebar.dart [hunk 6]
```diff
@@ -34,6 +37,20 @@ class Sidebar {
   /// Specifies whether the [Sidebar] can be resized by dragging or not.
   final bool? isResizable;
 
+  /// If true, the sidebar will close when dragged below [minWidth]. Use
+  /// [dragClosedBuffer] configure how far below [minWidth] it needs to be
+  /// dragged to trigger this behavior. Defaults to true.
```
Comment (GroovinChip) [inline lib/src/layout/sidebar.dart]:
```suggestion
  /// dragged to trigger this behavior. 
  /// 
  /// Defaults to `true`.
```

File: lib/src/layout/sidebar.dart [hunk 7]
```diff
@@ -34,6 +37,20 @@ class Sidebar {
...
+  final bool dragClosed;
+
+  /// If [dragClosed] is true, the sidebar will be hidden when dragged this far
+  /// below [minWidth].  Defaults to half of [minWidth]. Set to 0 to cause the
```
Comment (GroovinChip) [inline lib/src/layout/sidebar.dart:48]:
```suggestion
  /// below [minWidth]. Defaults to half of [minWidth]. Set to `0` to cause the
```

File: lib/src/layout/window.dart [hunk 8]
```diff
@@ -201,18 +202,39 @@ class _MacosWindowState extends State<MacosWindow> {
                 height: height,
                 child: GestureDetector(
                   behavior: HitTestBehavior.opaque,
+                  onHorizontalDragStart: (_) {
```
Comment (GroovinChip) [inline lib/src/layout/window.dart]: Please add a test for this behavior

File: lib/src/layout/window.dart [hunk 9]
```diff
@@ -201,18 +202,39 @@ class _MacosWindowState extends State<MacosWindow> {
...
+                    _sidebarDragStartWidth = _sidebarWidth;
+                  },
                   onHorizontalDragUpdate: (details) {
+                    final sidebar = widget.sidebar!;
```
Comment (GroovinChip) [inline lib/src/layout/window.dart:211]: Please add a test for this behavior

File: lib/src/layout/window.dart [hunk 10]
```diff
@@ -201,18 +202,39 @@ class _MacosWindowState extends State<MacosWindow> {
...
                     setState(() {
+                      var newWidth =
+                          _sidebarDragStartWidth + details.localPosition.dx;
```
Comment (GroovinChip) [inline lib/src/layout/window.dart]: Please add a test for this behavior

File: lib/src/layout/window.dart [hunk 11]
```diff
@@ -201,18 +202,39 @@ class _MacosWindowState extends State<MacosWindow> {
...
+
+                      if (sidebar.startWidth != null &&
+                          sidebar.snapToStartBuffer != null &&
+                          (newWidth - sidebar.startWidth!).abs() <=
+                              sidebar.snapToStartBuffer!) {
+                        newWidth = sidebar.startWidth!;
```
Comment (GroovinChip) [inline lib/src/layout/window.dart:221]: Please add a test for this behavior

File: lib/src/layout/window.dart [hunk 12]
```diff
@@ -201,18 +202,39 @@ class _MacosWindowState extends State<MacosWindow> {
...
+                      }
+
+                      if (sidebar.dragClosed) {
```
Comment (GroovinChip) [inline lib/src/layout/window.dart:224]: Please add a test for this behavior

File: lib/src/layout/window.dart [hunk 13]
```diff
@@ -201,18 +202,39 @@ class _MacosWindowState extends State<MacosWindow> {
...
+                        final closeBelow =
+                            sidebar.minWidth - sidebar.dragClosedBuffer;
+                        _showSidebar = newWidth >= closeBelow;
```
Comment (GroovinChip) [inline lib/src/layout/window.dart:227]: Please add a test for this behavior

File: lib/src/layout/window.dart [hunk 14]
```diff
@@ -201,18 +202,39 @@ class _MacosWindowState extends State<MacosWindow> {
...
+                      }
+
                       _sidebarWidth = math.max(
-                        widget.sidebar!.minWidth,
+                        sidebar.minWidth,
```
Comment (GroovinChip) [inline lib/src/layout/window.dart:231]: Please add a test for this behavior

File: lib/src/layout/window.dart [hunk 15]
```diff
@@ -201,18 +202,39 @@ class _MacosWindowState extends State<MacosWindow> {
...
                         math.min(
-                          math.min(widget.sidebar!.maxWidth!, width),
-                          _sidebarWidth + details.delta.dx,
+                          math.min(sidebar.maxWidth!, width),
```
Comment (GroovinChip) [inline lib/src/layout/window.dart]: Please add a test for this behavior

File: lib/src/layout/window.dart [hunk 16]
```diff
@@ -201,18 +202,39 @@ class _MacosWindowState extends State<MacosWindow> {
...
+                          newWidth,
                         ),
                       );
-                      if (_sidebarWidth == widget.sidebar!.minWidth)
+
+                      if (_sidebarWidth == sidebar.minWidth)
```
Comment (GroovinChip) [inline lib/src/layout/window.dart:238]: Please add a test for this behavior

File: lib/src/layout/window.dart [hunk 17]
```diff
@@ -201,18 +202,39 @@ class _MacosWindowState extends State<MacosWindow> {
...
                         _sidebarCursor = SystemMouseCursors.resizeRight;
-                      else if (_sidebarWidth == widget.sidebar!.maxWidth)
+                      else if (_sidebarWidth == sidebar.maxWidth)
```
Comment (GroovinChip) [inline lib/src/layout/window.dart:240]: Please add a test for this behavior

File: test/layout/window_test.dart [hunk 18]
```diff
@@ -38,12 +38,9 @@ void main() {
...
 
-      final sidebarFinder = find.byKey(
-        const ValueKey('sidebar'),
-        skipOffstage: true,
-      );
-      final resizerFinder = find.byKey(const ValueKey('sidebarResizer'));
-      final backgroundFinder = find.byKey(const ValueKey('background'));
+      final sidebarFinder = find.byType(AnimatedPositioned).first;
+      final resizerFinder = find.byType(AnimatedPositioned).at(3);
+      final backgroundFinder = find.byType(AnimatedPositioned).at(1);
```
Comment (GroovinChip) [inline test/layout/window_test.dart:43]: @jmatth are you doing anything with these finders in your test? Simply finding them by type is not really a test. Or is your work on improving the tests not complete yet?

File: CHANGELOG.md [hunk 19]
```diff
@@ -1,3 +1,8 @@
+## [0.10.3]
```
Comment (GroovinChip) [inline CHANGELOG.md]:
```suggestion
## [0.12.1]
```

Comment (GroovinChip) [reply CHANGELOG.md]: Need to change this in `pubspec.yaml` as well.

File: test/layout/window_test.dart (hunk 18 above)
Comment (jmatth) [reply test/layout/window_test.dart:43]: They're used in like every test in that file:

```bash
$ rg resizerFinder window_test.dart
42:      final resizerFinder = find.byType(AnimatedPositioned).at(3);
83:        await tester.drag(resizerFinder, const Offset(safeDelta, 0));
92:        await tester.drag(resizerFinder, const Offset(overflowDelta, 0));
102:            await tester.startGesture(tester.getCenter(resizerFinder));
114:        await tester.drag(resizerFinder, const Offset(-safeDelta, 0));
129:              resizerFinder,
145:            await tester.drag(resizerFinder, const Offset(-overflowDelta, 0));
158:                await tester.startGesture(tester.getCenter(resizerFinder));
182:              await tester.startGesture(tester.getCenter(resizerFinder));
198:            await tester.drag(resizerFinder, const Offset(-overflowDelta, 0));
209:              await tester.startGesture(tester.getCenter(resizerFinder));
233:                await tester.startGesture(tester.getCenter(resizerFinder));
259:            await tester.drag(resizerFinder, Offset(snapToStartBuffer * 2, 0));
268:              resizerFinder,
```

`sidebarFinder` and `backgroundFinder` are both called from inside the helper `expectSidebarOpen`, which is then used in every test:

```bash
$ rg expectSidebarOpen window_test.dart
45:      final expectSidebarOpen = (tester, {required double width}) {
72:        expectSidebarOpen(tester, width: startWidth);
86:        expectSidebarOpen(tester, width: startWidth + safeDelta);
95:        expectSidebarOpen(tester, width: maxWidth);
108:        expectSidebarOpen(tester, width: maxWidth);
117:        expectSidebarOpen(tester, width: startWidth - safeDelta);
137:            expectSidebarOpen(tester, width: minWidth);
174:            expectSidebarOpen(tester, width: startWidth - safeDelta);
201:            expectSidebarOpen(tester, width: minWidth);
215:          expectSidebarOpen(tester, width: minWidth);
237:            expectSidebarOpen(tester, width: startWidth);
242:            expectSidebarOpen(
262:            expectSidebarOpen(
272:            expectSidebarOpen(tester, width: startWidth);
```

**Comment by GroovinChip** [comment]
@jmatth can you address the package analysis check please? I can re-review after that.

**Comment by jmatth** [comment]
Heh, that `new` keyword it's complaining about is in a docstring in a file I didn't even modify. The Flutter `stable` channel probably updated its default analysis rules. But whatever, fixed it @GroovinChip

**Review by GroovinChip — CHANGES_REQUESTED** [review]
@jmatth  everything looks good, there's just one thing left to do - update the version in `pubspec.yaml` to match the version in `CHANGELOG.md`.

**Review by GroovinChip — APPROVED** [review]
LGTM, waiting on checks

**Pull request merged**